pgjdbc/src/main/java/org/postgresql/jdbc/TimestampUtils.java

```diff
@@ -1514,27 +1520,41 @@ private static long toPgSecs(long secs) {
   /**
    * Converts the SQL Date to binary representation for {@link Oid#DATE}.
    *
-   * @param tz The timezone used.
+   * @param tz The timezone used. If <code>null</code>, then no timezone offset will be applied to the date.
    * @param bytes The binary encoded date value.
    * @param value value
    * @throws PSQLException If binary format could not be parsed.
    */
   public void toBinDate(@Nullable TimeZone tz, byte[] bytes, Date value) throws PSQLException {
     long millis = value.getTime();
-
+    boolean isBce = false;
     if (tz == null) {
       tz = getDefaultTz();
     }
-    // It "getOffset" is UNTESTED
-    // See org.postgresql.jdbc.AbstractJdbc2Statement.setDate(int, java.sql.Date,
-    // java.util.Calendar)
-    // The problem is we typically do not know for sure what is the exact required date/timestamp
-    // type
-    // Thus pgjdbc sticks to text transfer.
-    millis += tz.getOffset(millis);
+    if (millis < GREGORIAN_CUTOVER) {
+      BaseCalendar calendar = getJulianCalendar();
+      CalendarDate calendarDate = calendar.getCalendarDate(millis, tz);
+      isBce = calendarDate.getEra().getAbbreviation().equals("B.C.E.");
+    }
 
-    long secs = toPgSecs(millis / 1000);
-    ByteConverter.int4(bytes, 0, (int) (secs / 86400));
+    // LocalDate does not work properly for BC dates, as the era is not carried over to the local
+    // date.
+    LocalDate localDate = value.toLocalDate();
+    if (isBce) {
+      // LocalDate accepts year 0 (so 0000-01-01 is a valid LocalDate). PostgreSQL expects dates to
+      // have a year value >= 1 (BC or BCE). So for BC we need to flip the sign of the
+      // year value and add one to get the right number of days since PostgreSQL epoch.
+      localDate = localDate.withYear(-localDate.getYear() + 1);
+    }
+    long epochDay = localDate.toEpochDay();
+    ByteConverter.int4(bytes, 0, (int) (epochDay - PG_EPOCH_DIFF.toDays()));
```

Comment:
@davecramer , I wonder if we need to fix `BC dates` in the current `master` branch for `LocalDate`

Reply:
Probably worth doing, yes. I can hold off releasing if you want to cobble together a PR ?

Reply:
Frankly speaking, I don't fully understand it. It looks like "LocalDate is incompatible with PGs handing of BC/BCE dates". If that is really the case, I am afraid it might mean https://github.com/pgjdbc/pgjdbc/pull/2464/files#diff-21e3023c84d9d5fb4f5f04678a094fce6ab5af36a5e73fdfaf495646f3b720d5R1283 introduced a BC/BCE error.

I'm fine if "binary handling for Oid.DATE" is postponed, however, I would like to figure out if there's BC/BCE bug introduced in 2464

Reply:
I may try to see if that is the case, likely Tues before I get to it though

Reply:
I just noticed my comment might be a little unclear: `LocalDate` works with BCE dates, but with the caveat that it accepts year 0 as a valid year, which is different from PostgreSQL. `java.util.Date#toLocalDate()` does __not__ work properly with BCE dates. If you for example have a `java.util.Date` in the year 100BCE, `java.util.Date#toLocalDate()`  will return a CE `LocalDate`.

Reply:
`java.util.Date#toLocalDate()` does not exist.
`java.sql.Date#toLocalDate` is not yet used in pgjdbc.

Reply:
@vlsi Any update on this? Is there any actions you would like me to take?

Reply:
So I am looking at this and I don't believe that the tests actually test binary mode. The data is not being transferred in binary AFAICT

Reply:
@vlsi yes, BC dates are broken in HEAD in TEXT mode

Reply:
> So I am looking at this and I don't believe that the tests actually test binary mode. The data is not being transferred in binary AFAICT

@davecramer I've just started the [DateTest.testSetDate()](https://github.com/olavloite/pgjdbc/blob/8e0c699a9b7b9662f6787323e6dfe29fa1a480dd/pgjdbc/src/test/java/org/postgresql/test/jdbc2/DateTest.java#L100) in debug mode and [set a breakpoint here](https://github.com/pgjdbc/pgjdbc/blob/aa1f58699d3ee0d8a149be60da44b756e1e618f2/pgjdbc/src/main/java/org/postgresql/jdbc/PgPreparedStatement.java#L1321) to be absolutely sure that they are sent in binary format. The test enters this branch and encodes the date in binary format.

Is there a specific reason why you think they are not encoded in binary format?

Reply:
I might have been a little vague about that. What I meant to say is that retrieving the data is not done in binary mode. So we aren't testing binary in testGetDate() 

In order to force binary you need to add PGProperty.PREPARE_THRESHOLD.set(props, -1); before the connection is created.

Reply:
Regarding binary transfer of dates from the backend to the frontend: That is correct, these tests don't do that. That is also not part of this change. This is purely for adding the possibility of sending date parameters in binary format from the frontend to the backend.

Currently on HEAD you can create a configuration that should send date parameters in binary format to the backend by adding the following to your connection properties: `properties.setProperty(PGProperty.BINARY_TRANSFER_ENABLE.getName(), String.format("%d", Oid.DATE))`.

That option is however ignored, as it is specifically removed by the driver at a later point. This change only changes that behavior. I'm happy to add additional tests for sending dates in binary format from the backend to the frontend, but as mentioned above; this PR does not introduce any changes in the current behavior of sending data from the backend to the frontend.

Reply:
> Frankly speaking, I don't fully understand it. It looks like "LocalDate is incompatible with PGs handing of BC/BCE dates". If that is really the case, I am afraid it might mean https://github.com/pgjdbc/pgjdbc/pull/2464/files#diff-21e3023c84d9d5fb4f5f04678a094fce6ab5af36a5e73fdfaf495646f3b720d5R1283 introduced a BC/BCE error.
> 
> I'm fine if "binary handling for Oid.DATE" is postponed, however, I would like to figure out if there's BC/BCE bug introduced in 2464

Hi @vlsi, 

as being the original author of #2464, I tried to follow the discussion here, but I have some problem to understand about which conversion actually fails. @davecramer later said that only text transfer from progresql to pgjdbc is broken:

> @vlsi yes, BC dates are broken in HEAD in TEXT mode

He then pinged me in another pull request about `OffsetTime` (which is always send in prepared statements as text, too, like all java.time types currently) because of a change to text parsing. I think the issue we see here just comes from the fact how we convert from Postgres TEXT format to LocalDate?

To help here and fix the recent changes done by me it would be good to get a failing test with get/setObject(LocalDate).

I know that in #2464 theres no test for pre-year-1 dates, so we should possibly add one and test it in binary and text mode.

Reply:
@uschindler apologies, I should have been more clear.

stmt.executeUpdate(TestUtil.insertSQL(table, "'0101-01-01 BC'")));
assertEquals(1, stmt.executeUpdate(TestUtil.insertSQL(table, "'0001-01-01'")));

Reading this back now gives us 0102-12-31 if I remember correctly.

Thanks

Reply:
If reading back as `LocalDate` or as `java.sql.Date`?

Reply:
This test seems to work: https://github.com/pgjdbc/pgjdbc/blob/master/pgjdbc/src/test/java/org/postgresql/test/jdbc2/DateTest.java#L68-L71

So theres no problem in parsing. I can make a similar test for `LocalDate` and verify that it works both in binary and text mode?

Reply:
Inserting into a timestamp with timezone and reading back as a date.
```
try ( Connection con = DriverManager.getConnection(url, props)) {
            Statement stmt = con.createStatement();
            stmt.execute("create table if not exists testtz (dt timestamptz)");
            stmt.execute("truncate testtz");
            assertEquals(1, stmt.executeUpdate("insert into testtz values ('0101-01-01 BC')" ));
            assertEquals(1, stmt.executeUpdate("insert into testtz values ('0001-01-01')" ));
            assertEquals(1, stmt.executeUpdate("insert into testtz values ('0001-01-01 BC')" ));
            assertEquals(1, stmt.executeUpdate("insert into testtz values ('0001-12-31 BC')" ));

            try ( ResultSet rs = stmt.executeQuery("select dt from testtz")) {
                Date d = null;
                assertTrue(rs.next());
                d = rs.getDate(1);
                assertNotNull(d);
                //0101-01-01 BC
                assertEquals(makeDate(-100, 1, 1), d);
                assertTrue(rs.next());
                d = rs.getDate(1);
                assertNotNull(d);
                // 0001-01-01
                assertEquals(makeDate(1, 1, 1), d);

                assertTrue(rs.next());
                d = rs.getDate(1);
                assertNotNull(d);
                // 0001-01-01 BC
                assertEquals(makeDate(0, 1, 1), d);

                assertTrue(rs.next());
                d = rs.getDate(1);
                assertNotNull(d);
                // 0001-12-31 BC
                assertEquals(makeDate(0, 12, 31), d);

                assertTrue(!rs.next());
            }

        } catch ( Exception ex ) {
            ex.printStackTrace();
        }
    }
    private static java.sql.Date makeDate(int y, int m, int d) {
        return new java.sql.Date(y - 1900, m - 1, d);
    }
```

Reply:
OK, I made a quick test for `LocalDate` and `LocalDateTime`. It returns one year difference for BC dates. I have no idea how this was before my changes, because there was no test at all.

What I figured out is the following: It depends on how you write your test!!!! Which is some strange thing in the LocalDate implementation in Java.time. If you do use a negative year the result is different than when you create a year with a positive date and set the ERA to BC.

So it looks like the output is correct.

All tests I wrote return correct dates also for BC, if you build them as expected. This shows the problem:

```
      LocalDate expected = LocalDate.of(1582, 9, 30)
          .with(ChronoField.ERA, IsoEra.BCE.getValue());
      System.err.println(expected);
```
 This returns '-1581-09-30' - BUMMER!

So I don't think theres a problem in my code it is just the way how the text representation looks like. There is also an existing test that exactly tests this with LocalDateTime in the same way (and it was preexistent).

Reply:
If we have a bug here and it should return '-1582-09-30' then we need to fix the whole parsing like done above (also for binaries). This is completely unrelated to the code changes in #2464 and #2467 and looks like more a general problem. 

This is also a problem in the previous impl. The PR here tries to workaround with this:
https://github.com/pgjdbc/pgjdbc/pull/2476/files#diff-21e3023c84d9d5fb4f5f04678a094fce6ab5af36a5e73fdfaf495646f3b720d5R1544-R1547

The problems appear here, because this only fixes the binary representation, but not the text representation, this is why we see differences.

Reply:
> Inserting into a timestamp with timezone and reading back as a date.
> 
> ```
> try ( Connection con = DriverManager.getConnection(url, props)) {
>             Statement stmt = con.createStatement();
>             stmt.execute("create table if not exists testtz (dt timestamptz)");
>             stmt.execute("truncate testtz");
>             assertEquals(1, stmt.executeUpdate("insert into testtz values ('0101-01-01 BC')" ));
>             assertEquals(1, stmt.executeUpdate("insert into testtz values ('0001-01-01')" ));
> ```

This looks like an unrelated issue to this PR. OK, so we have multiple problems here:
- database date or timestamp -> LocalDate for BC dates seem to are one year off (see my previous comment). This is an preexisting issue
- database timestamptz seems to be broken if you get them as dates (in text mode)

I have the feeling the latter may be caused by my code, but that may be a problem of timezones. The server applies the local timezone to the result (timestamptz) and then parsing breaks, because the timezone is applied and shifting the values. That's the problem of the f*cking timestamptz datatype! I try to write a test.

Reply:
> OK, I made a quick test for `LocalDate` and `LocalDateTime`. It returns one year difference for BC dates. I have no idea how this was before my changes, because there was no test at all.
> 
> What I figured out is the following: It depends on how you write your test!!!! Which is some strange thing in the LocalDate implementation in Java.time. If you do use a negative year the result is different than when you create a year with a positive date and set the ERA to BC.
> 
> So it looks like the output is correct.
> 
> All tests I wrote return correct dates also for BC, if you build them as expected. This shows the problem:
> 
> ```
>       LocalDate expected = LocalDate.of(1582, 9, 30)
>           .with(ChronoField.ERA, IsoEra.BCE.getValue());
>       System.err.println(expected);
> ```
> 
> This returns '-1581-09-30' - BUMMER!
> 
> So I don't think theres a problem in my code it is just the way how the text representation looks like. There is also an existing test that exactly tests this with LocalDateTime in the same way (and it was preexistent).

I wrote a test for `LocalDate` similar to the pre-existing test:

  @Test
  public void testBcDate() throws SQLException {
    try (Statement stmt = con.createStatement(); ResultSet rs = stmt.executeQuery("SELECT '1582-09-30 BC'::date")) {
      assertTrue(rs.next());
      LocalDate expected = LocalDate.of(1582, 9, 30)
          .with(ChronoField.ERA, IsoEra.BCE.getValue());
      LocalDate actual = rs.getObject(1, LocalDate.class);
      assertEquals(expected, actual);
      assertFalse(rs.next());
    }
  }


Works. But keep in mind: like the original text, the date as converted to BCE is numerically one year off if printed as ISO timestamp using toString(). This has to do with the different meaning of "u" and "y" in date formatters. ISO uses "u" which is the year, possibly negative, while "y" is "year of era" (only positive). So catually the workarounds done here are plain wrong, because they are inconsistent with how pgjdbc works at other places.

Reply:
What @davecramer found out is a different issue. I will look into this.

Should I provide a PR to add the above test to GetObject310Test?

Reply:
Yes, please!

Reply:
One important this and this is also what I tried to fix in my previous 2 pull requests: Never mix old and new Java time API! When you read/write a timestamp or date from wire format always use java.time *OR* legacy Clendar APIs. The code here mixes both to work around a bug, but that should never be done. If you want to support setting java.sql.Date on a prepared statement only use legacy Java API. Never ever use java.sql.Date#toLocalDate() - like done here, this brings you into big trouble.

I hope Java at some point deprecates the legacy APIs!

Reply:
> Yes, please!

Here is a PR: #2488

Feel free to commit tests to it. I started with the first test that makes sure that `ResultSet#getObject(LocalDate)` works with BC dates.

What seems to be @davecramer 's issue is that there seem to be some problems when you have a timestamptz in datbase and you read it as `java.sql.Date`. This issue may be caused by #2464 or #2467, but could also be preexisting (I will try to revert my two commits after I have a test).

Reply:
Hi @vlsi 

> Frankly speaking, I don't fully understand it. It looks like "LocalDate is incompatible with PGs handing of BC/BCE dates". If that is really the case, I am afraid it might mean https://github.com/pgjdbc/pgjdbc/pull/2464/files#diff-21e3023c84d9d5fb4f5f04678a094fce6ab5af36a5e73fdfaf495646f3b720d5R1283 introduced a BC/BCE error.
> 
> I'm fine if "binary handling for Oid.DATE" is postponed, however, I would like to figure out if there's BC/BCE bug introduced in 2464

I can explain that: It is all fine in master (and also before) with LocalDate. In our code we never use the year setter on localDate because this one is ISO (so ISO year 0 is -1 BC). If you have a BC year, you need to construct your date in a different way: at all places in our code we use the `ChronoUnit.YEAR_OF_ERA` (https://docs.oracle.com/javase/8/docs/api/java/time/temporal/ChronoField.html#YEAR_OF_ERA) setter when we parse strings. The bug here is (a well known bug) is that 'java.sql.Date#toLocalDate()' does use the year setter so the patching is required. But this is only done as workaround for the problems in that buggy method (it should be placed in forbiddenapis signature file!).

To fix this see my comment on main thread.

The issue that @davecramer found is a different one with text representation and was indeed introduced by me. It is fixed in the other PR that moves over the good tests from this PR to a separate commit, so we can release pgjdbc with my bugfix. In fact the problem was the place in the code that you were complaining for speed reasons (`TimeZone.get(ZoneOffset)` vs. `SimpleTimeZone`). Actually when using Calendars you should use simple timezone legacy class, too as the ZoneOffset timezones sometimes break with buggy timezones sometimes sent by server for BC dates.

Reply:
Here is the fix for @davecramer's issue in his Toronto timezone where backing postgres server does return crazy offsets: https://github.com/pgjdbc/pgjdbc/pull/2488/files#diff-21e3023c84d9d5fb4f5f04678a094fce6ab5af36a5e73fdfaf495646f3b720d5R152-R157

Reply:
In fact, we should maybe change our code at other places to use `https://docs.oracle.com/javase/9/docs/api/java/time/chrono/IsoChronology.html#date-java.time.chrono.Era-int-int-int-` when we construct LocalDate from text strings.

I can make a separate PR for this.